src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs

```diff
@@ -3125,20 +3144,752 @@ End Sub
+Imports System.Collections.Generic
+Imports System.Linq
+
+Namespace Ns
+    Public Class Hoo
+        Public Sub Goo(j As IEnumerable(Of Integer))
+            Dim count = 0
+            If j.TryGetNonEnumeratedCount(count) then
+
+            End If
+            j.ElementAt(10)
+        End Sub
+    End Class
+End Namespace
+";
+            await VerifyVisualBasicAsync(vbCode);
+        }
+
+        [Fact]
+        public async Task TestByDefaultSupposeMethodEnuemrationArgument()
+        {
+            var csharpCode = @"
+using System.Collections.Generic;
+using System.Linq;
+
+public class Bar
+{
+    public void Sub(IEnumerable<int> j, IEnumerable<int> i)
+    {
+        [|j|].UnionBy(i, p => p).ElementAt(10);
+        Method([|j|]);
+    }
+
+    private void Method(IEnumerable<int> k)
+    {
+        foreach (var p in k) { }
+    }
+}";
+            await VerifyCSharpAsync(csharpCode);
+
+            var vbCode = @"
+Imports System.Collections.Generic
+Imports System.Linq
+
+Namespace Ns
+    Public Class Hoo
+        Public Sub Goo(j As IEnumerable(Of Integer), i As IEnumerable(Of Integer))
+            [|j|].UnionBy(i, Function(p) p).ElementAt(10)
+            Method([|j|])
+        End Sub
+
+        Private Sub Method(j As IEnumerable(Of Integer))
+            For Each kk in j
+            Next
+        End Sub
+    End Class
+End Namespace
+";
+            await VerifyVisualBasicAsync(vbCode);
+        }
+
+        [Fact]
+        public async Task TestNoEnumerationMethodFromEditorConfig()
+        {
+            var csharpCode = @"
+using System.Collections.Generic;
+using System.Linq;
+
+public class Bar
+{
+    public void Sub(IEnumerable<int> j, IEnumerable<int> i)
+    {
+        Method1(j);
+        j.UnionBy(i, p => p).ElementAt(10);
+    }
+
+    private void Method1(IEnumerable<int> k)
+    {
+    }
+}";
+            await VerifyCSharpAsync(csharpCode, "M:Bar.Method1*");
+
+            var vbCode = @"
+Imports System.Collections.Generic
+Imports System.Linq
+
+Namespace Ns
+    Public Class Hoo
+        Public Sub Goo(j As IEnumerable(Of Integer), i As IEnumerable(Of Integer))
+            Method1(j)
+            j.UnionBy(i, Function(p) p).ElementAt(10)
+        End Sub
+
+        Private Sub Method1(j As IEnumerable(Of Integer))
+        End Sub
+    End Class
+End Namespace
+";
+            await VerifyVisualBasicAsync(vbCode, "M:Ns.Hoo.Method1*");
+        }
+
+        [Fact]
+        public async Task TestLinqMethodFromEditorConfig()
+        {
+            var csharpCode = @"
+using System.Collections.Generic;
+using System.Linq;
+
+public class Bar
+{
+    public void Sub(IEnumerable<int> j, IEnumerable<int> i)
+    {
+        [|j|].UnionBy(i, p => p).ElementAt(10);
+        Chain([|j|]).ElementAt(100);
+    }
+
+    private IEnumerable<int> Chain(IEnumerable<int> k)
+    {
+        return k;
+    }
+}";
+            await VerifyCSharpAsync(csharpCode, customizedLinqChainMethods: "M:Bar.Chain*");
```

Comment:
I am not understanding this config well, what would be changed in the diagnostics if the `Chain` method was not configured here?

Reply:
The diagnostic would be the same, but the analyzer would think it is `Chain` enumerated the `j`, not `ElementAt`

Reply:
So the diagnostic position would be different? Like `Chain(j).[|ElementAt(100);|]` or so?

Reply:
Extension methods can also be invoked in this way
```
var i = Enumerable.Range(1, 10);
var a = Linq.Select(i, j => j + 1).ElementAt(100);
```
Comparing with the normal way
```
var i = Enumerable.Range(1, 10);
var a = i.Select(j => j + 1).ElementAt(100);
```
I would like to keep the diagnostic reported under the local reference or parameter reference location to maintain the same behavior in all of the diagnostics.

Reply:
I agree with @buyaa-n in the sense that we should have a test such that diagnostic is reported in presence of editorconfig setting and no diagnostic is reported without the setting, that makes the test clear to understand. Also make this test have a parameterized `string editorconfig` so we can have `var arg = editorconfig.IsEmpty ? "j" : "[|j|]";` and just substitute `arg` in interpolated string for the test code. This matches how we write all of our editorconfig based tests in this repo.

Another way would be: https://github.com/dotnet/roslyn-analyzers/blob/89942866e0fde9a71327be082183b6b3b972a08b/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/CancellationTokenParametersMustComeLastTests.cs#L561-L564

Reply:
BTW I forget to update one thing about this thread.
> The diagnostic would be the same, but the analyzer would think it is Chain enumerated the j, not ElementAt

This comment is made when by default, all the methods are thought would enumerate the parameter.
And Linq chain is also fitting into this pattern, it would enumerate the parameter.
However, right now, if there is no configuration there would be no diagnostics.

Reply:
I have update the tests to both test with/without editor config